src/OrchardCore/OrchardCore.Abstractions/Modules/ModularApplicationContext.cs

```diff
@@ -48,12 +49,17 @@ private void EnsureInitialized()
 
         private IEnumerable<Module> GetModules()
         {
-            var modules = new List<Module>();
+            var modules = new ConcurrentBag<Module>();
             modules.Add(new Module(_environment.ApplicationName, true));
 
             foreach (var provider in _moduleNamesProviders)
             {
-                modules.AddRange(provider.GetModuleNames().Select(name => new Module(name, name == _environment.ApplicationName)));
+                var names = provider.GetModuleNames().Distinct();
+
+                Parallel.ForEach(names, new ParallelOptions { MaxDegreeOfParallelism = 8 }, (name) =>
+                {
+                    modules.Add(new Module(name, name == _environment.ApplicationName));
```

Comment:
I am wondering... should name and application name be case insensitive? :/

Reply:
I actually think this whole section is kinda wrong... shouldnt it be more like this...

```
	var modules = _moduleNamesProviders
		.SelectMany(x => x.GetModuleNames())
		.Distinct()
		.AsParallel()
		.Select(name => new Module(name, name == _environment.ApplicationName));
```